### ISSUE TITLE
List tables

### ISSUE THREAD

**Issue by jardakotesovec** [knex/knex#360]
I did not find function that would allow me to get list of tables in database. And it is not obvious to me from doc if its possible to list all columns with 'columnInfo' function.

**Comment by whitelynx** [comment]
For PostgreSQL, we should probably be explicit about the schema, since it's possible to configure it such that `pg_catalog` is not in the search path:
```sql
SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname='public'
```

**Comment by whitelynx** [comment]
Also, this page has a good summary of the ways of querying available databases, schemas, tables, etc. in various databases (though it's missing sqlite): http://troels.arvin.dk/db/rdbms/#cli-list_of_databases

**Comment by samgranger** [comment]
+1 on this

**Comment by rex** [comment]
Big +1 from me! :+1:

**Comment by zachelrath** [comment]
+1

**Comment by samhagan** [comment]
+1

**Comment by NikolaBorislavovHristov** [comment]
+1

**Comment by tgriesser** [comment]
Ok, let's add this as `knex.listTables()`, adding to the 0.14 project

**Comment by RWOverdijk** [comment]
:+1: This is very useful. :)

**Comment by zh99998** [comment]
it's 0.13.0 now but i still can't find it in document

**Comment by elhigu** [comment]
@zh99998 nobody has implemented it yet.

**Comment by verekia** [comment]
For SQLite, you can run this query:
```js
knex('sqlite_master').where('type', 'table')
```
To get an array of just the names inside an `async` function:
```js
(await knex('sqlite_master').where('type', 'table')).map(x => x.name)
```

There might be equivalent metadata tables in other database types, but at least this works well for SQLite and testing things in memory.

**Comment by GeoffreyPlitt** [comment]
Still can't figure out how to do this for Postgres, can we get a working example? knex.raw() doesn't seem to execute SQL, it has to be passed to something else?

**Comment by elhigu** [comment]
For example all these should execute knex.raw query just fine: `knex.raw(...).then(result => ...)` or `result = await knex.raw(...)` or `Promise.resolve(knex.raw(...)).then(result => ...)` or `otherPromise.then(() => knex.raw(...)).then(result => ...)`.

@GeoffreyPlitt do you have some specific case that does not work?

**Comment by GeoffreyPlitt** [comment]
You're right. I was confused by the results being an instance of class Result and couldn't figure out if the query had run or not. I find this returned simpler data:

```
knex('pg_catalog.pg_tables')
.select('tablename')
.where({schemaname:'public'})
```

**Comment by codeclown** [comment]
Here's a PR: https://github.com/tgriesser/knex/pull/2560

**Comment by chrislondon** [comment]
For anyone interested I'm currently using this method.  I've only tested it in PostgreSQL (the database I use) so if anyone sees any issues let me know.

`export function listTables(knex:Knex) {
    let query:string;
    let bindings: string[];

    switch(knex.client.constructor.name) {
        case 'Client_MSSQL':
            query = 'SELECT table_name FROM information_schema.tables WHERE table_schema = \'public\' AND table_catalog = ?',
            bindings = [ knex.client.database() ];
            break;
        case 'Client_MySQL':
        case 'Client_MySQL2':
            query = 'SELECT table_name FROM information_schema.tables WHERE table_schema = ?';
            bindings = [ knex.client.database() ];
            break;
        case 'Client_Oracle':
        case 'Client_Oracledb':
            query = 'SELECT table_name FROM user_tables';
            break;
        case 'Client_PG':
            query =  'SELECT table_name FROM information_schema.tables WHERE table_schema = current_schema() AND table_catalog = ?';
            bindings = [ knex.client.database() ];
            break;
        case 'Client_SQLite3':
            query = "SELECT name AS table_name FROM sqlite_master WHERE type='table'";
            break;
    }

    return knex.raw(query, bindings).then(function(results) {
        return results.rows.map((row) => row.table_name);
    });
}`

**Comment by notaphplover** [comment]
> For anyone interested I'm currently using this method. I've only tested it in PostgreSQL (the database I use) so if anyone sees any issues let me know.
> 
> ```
> export function listTables(knex:Knex) {
>     let query:string;
>     let bindings: string[];
> 
>     switch(knex.client.constructor.name) {
>         case 'Client_MSSQL':
>             query = 'SELECT table_name FROM information_schema.tables WHERE table_schema = \'public\' AND table_catalog = ?',
>             bindings = [ knex.client.database() ];
>             break;
>         case 'Client_MySQL':
>         case 'Client_MySQL2':
>             query = 'SELECT table_name FROM information_schema.tables WHERE table_schema = ?';
>             bindings = [ knex.client.database() ];
>             break;
>         case 'Client_Oracle':
>         case 'Client_Oracledb':
>             query = 'SELECT table_name FROM user_tables';
>             break;
>         case 'Client_PG':
>             query =  'SELECT table_name FROM information_schema.tables WHERE table_schema = current_schema() AND table_catalog = ?';
>             bindings = [ knex.client.database() ];
>             break;
>         case 'Client_SQLite3':
>             query = "SELECT name AS table_name FROM sqlite_master WHERE type='table'";
>             break;
>     }
> 
>     return knex.raw(query, bindings).then(function(results) {
>         return results.rows.map((row) => row.table_name);
>     });
> }
> ```

@chrislondon Nice code, ¿could it be a good idea to switch using knex.client.driverName?

**Comment by ekcom** [comment]
If people still are finding this issue (like I did), see [knex-schema-inspector](https://github.com/knex/knex-schema-inspector).

**Comment by mercmobily** [comment]
@ekcom Are you still maintaining this module? If  you are, we might add it to the documentation and close this issue. Thank you for writing it!

**Comment by mercmobily** [comment]
Also keep an eye on this: https://github.com/knex/knex/issues/624
